Search Relevance Workbench Page (#9736)
opensearch-project/documentation-website: _search-plugins/search-relevance/optimize-hybrid-search.md

{
  "name": "hybrid_query_lexical",
  "query": "{\"query\":{\"hybrid\":{\"queries\":[{\"match\":{\"title\":\"%SearchText%\"}},{\"match\":{\"category\":\"%SearchText%\"}}]}}}",
  "index": "ecommerce"
}'
```

However, we assume hybrid search optimization to add most value when combining lexical and vector-based search results. The requirement for your search configuration is to contain a hybrid search query with two query clauses: one textual query clause and one neural query clause. The search pipeline that deals with the combination of results is handled by the hybrid search optimization process itself. Here is an example of how a suitable search configuration for a hybrid search optimization experiment might look like:

```json
PUT _plugins/_search_relevance/search_configurations
{
  "name": "hybrid_query_text",
  "query": "{\"query\":{\"hybrid\":{\"queries\":[{\"multi_match\":{\"query\":\"%SearchText%\",\"fields\":[\"id\",\"title\",\"category\",\"bullets\",\"description\",\"attrs.Brand\\\",\"attrs.Color\"]}},{\"neural\":{\"title_embedding\":{\"query_text\":\"%SearchText%\",\"k\":100,\"model_id\":\"lRFFb5cBHkapxdNcFFkP\"}}}]}},\"size\":10}",
  "index": "ecommerce"
}
```

> The specified model ID in the `query` has to be valid and point to a model available in OpenSearch. The corresponding field containing the embeddings for the neural search part (`title_embedding` in this example) has to exist in the target index prior to running this experiment. The [`search-relevance` repository](https://github.com/opensearch-project/search-relevance) contains an end-to-end example guiding you in case you need more information on how to get started and meet the requirements for this experiment type.

Reviewer: ```suggestion
The model ID specified in the `query` must be a valid model ID for a model deployed in OpenSearch. The target index must contain the field for neural search embeddings (in this example, `title_embedding`). 

For an end-to-end example, see the [`search-relevance` repository](https://github.com/opensearch-project/search-relevance).
```

## Running a hybrid search optimzation experiment

A hybrid search optimization experiment can be created with the experiments endpoint of the Search Relevance Workbench.

### Endpoint

```json
PUT _plugins/_search_relevance/experiments
```

### Example request

```json
PUT _plugins/_search_relevance/experiments
{
  "querySetId": "b16a6a2b-ed6e-49af-bb2b-fc739dcf24e6",
  "searchConfigurationList": ["508a8812-27c9-45fc-999a-05f859f9b210"],
  "judgmentList": ["1b944d40-e95a-43f6-9e92-9ce00f70de79"],
  "size": 10,
  "type": "HYBRID_OPTIMIZER"
}

> ```

Reviewer: Add copy-curl

### Example response

```json
{
  "experiment_id": "0f4eff05-fd14-4e85-ab5e-e8e484cdac73",
  "experiment_result": "CREATED"
}
```

## Experimentation process

The hybrid search optimization experiment runs different evaluations based on the search configuration. The following parameters and parameter values are taken into account:

* Two normalization techniques: `l2` and `min_max`.
* Three combination techniques: `arithmetic_mean`, `harmonic_mean`, `geometric_mean`.
* The lexical and neural search weights, which are values ranging from `0.0` to `1.0` in 0.1 increments.

> Every query in the query set will be executed for all different parameter combinations and the results evaluated by using the judgment list.

Reviewer: ```suggestion
Every query in the query set is executed for all different parameter combinations, and the results evaluated by using the judgment list.
```

## Evaluating the results

The results for each evaluation are stored and can be viewed in the frontend by selecting the corresponding experiment in the overview of past experiments:

<img src="{{site.url}}{{site.baseurl}}/images/search-relevance-workbench/experiment_overview_hybrid_search_optimization.png" alt="Compare search results"/>{: .img-fluid }

> This leads to the detailed view for all executed queries with their calculated search metrics:

Reviewer: ```suggestion
All executed queries with their calculated search metrics are displayed, as shown in the following image.
```

---
layout: default
title: Optimizing hybrid search
nav_order: 60
parent: Using Search Relevance Workbench
grand_parent: Search relevance
has_children: false
has_toc: false
---

# Optimizing hybrid search

When you use hybrid search in OpenSearch, a key question is how to combine results from lexical and vector-based search effectively. OpenSearch provides different techniques with various parameters you can experiment with to find the best setup for your application. What works best, however, depends heavily on your data, user behavior, and application domain—there is no one-size-fits-all solution.

The Search Relevance Workbench helps you systematically find the ideal set of parameters for your needs.

## Requirements

Internally, optimizing hybrid search involves running multiple search quality evaluation experiments. For these experiments, you need a query set, judgments, and a search configuration.

> Currently, the Search Relevance Workbench supports hybrid search optimization with up to two query clauses. While hybrid search typically combines vector and lexical queries, you can run hybrid search optimization with two lexical query clauses:

Reviewer 2: There must be exactly two sub-queries in hybrid query, in case it's only one sub-query there will be a validation error.
